backfiller/backfiller/main.py

```diff
@@ -0,0 +1,227 @@
+
+	Backfill from node/stream/variants to base_dir/stream/variants for each node in nodes. If nodes is None, use get_nodes() to get a list of nodes to backfill from. Passes hours, start, stop and order to backfill_node to control which hours are backfilled and in which order. By default all hours are backfilled. If backfilling from a node raises an exception, this just goes onto the next node."""
+
+	if nodes is None:
+		nodes = get_nodes()
+
+	#ideally do this in parallel
+	for node in nodes:
+		try:
+			backfill_node(base_dir, node, stream, variants, hours, start, stop, order=order)
+		except Exception as e:
+			print node, e
+
+
+def is_iterable(x):
+	"""Test whether input is iterable."""
+	try:
+ 		iter(x)
+	except TypeError:
+		return False
+	return True
+
+
+def backfill_node(base_dir, node, stream, variants, hours=None, start=None,
+		stop=None, order=None, recent_cutoff=60):
+	"""Backfill from remote node.
+
+	Backfill from node/stream/variants to base_dir/stream/variants.
+
+	Keyword arguments:
+	hours -- If None (default), backfill all available hours. If iterable, backfill only hours in iterable. Otherwise backfill the last N hours, starting with the lastest.
+	start -- Only backfill hours starting after or equal to this datetime object. If None (default), backfill all hours.
+	stop -- Only backfill hours starting before or equal to this datetime object. If None (default), backfill all hours.
+	order -- If 'random', randomise the order of hours. If 'forward', sort the hours in acceding order. If 'reverse', sort the hours in descending order. Otherwise, do not change the order of hours (default).
+	recent_cutoff -- Skip backfilling segments younger than this number of seconds to prioritise letting the downloader grab these segments."""
+
+	if hours is None:
+		hours = list_remote_hours(node, stream, variant)
+	elif is_iterable(hours):
+		None
+	else:
+		n_hours = hours
+		if n_hours < 1:
+			raise ValueError('Number of hours has to be 1 or greater')
+		now = datetime.datetime.utcnow()
+		hours = [(now - i * timedelta(hours=1)).strftime(HOUR_FMT) for i in range(n_hours)]
+
+	if start is not None:
+		hours = [hour for hour in hours if hour >= start]
+	if stop is not None:
+		hours = [hour for hour in hours if hour <= stop]
+
+	# useful if running in parallel and expecting to fetch a segments from
+	# multiple hours (say on start up) so that you don't try to backfill the
+	# same hour at the same time 
+	if order == 'random':
+		hours = random.shuffle(hours)
+	elif order == 'forward':
+		sort(hours)
+	elif order == 'reverse':
+		sort(hours, reverse=True)
+
+	for variant in variants:
+
+		for hour in hours:
+
+			local_segments = set(list_local_segments(base_dir, stream, variant, hour))
+			remote_segments = set(list_remote_segments(node, stream, variant, hour))
+			missing_segments = remote_segments - local_segments
+
+			for missing_segment in missing_segments:
+
+				#ignore temporary files
+				if 'temp' in missing_segment:
```

Comment:
this will cause any segment whose hash happens to contain the letters 'temp' to not be replicated.
You should probably use `common.parse_segment_path()` instead, and simply ignore any segments that don't parse correctly (gives a ValueError). This can also replace the `.ts` check and the segment time parsing below.

Reply:
Just for fun and because I thought this was amusing, I ran the numbers on the chance of this occuring:

Let's consider how many hashes contain "temp" as a substring.
Due to how base64 works, and that we're representing 256 bits in 43 characters,
the last character has a bit form of xxxx00, ie one of: AEIMQUYcgkosw048,
which does not include p. So last 4 chars can't be "temp", so start char
of the temp substring must be in the first 38 chars.
So there are 38 positions for "temp". For each one, the other 38 unrestricted chars + 1 final char can be anything, which is 64^38 * 16.
So there are 38 * 16 * 64^38 = (19 * 2) * 2^4 * (2^6)^38 = 19 * 2^233 hashes containing
the substring "temp". There are of course 2^256 hashes total, so the chance of a hash
containing "temp" is 19 * 2^233 / 2^256 = 19 / 2^23 or 0.00023%.
The chance of N hashes NOT containing "temp" in any of them is (1 - 19/2^23)^N.
If we expect a week's worth of segments at 6s per segment, that's 100800 segments,
giving us a chance of 79.59% of hitting this in practice.

Reply:
Actually it's a little less likely because i didn't consider things that contain "temp" twice, so eg. a hash like "temptemp000000..." would be double-counted.